src/libopensc/card-cac1.c

```diff
@@ -0,0 +1,555 @@
+
+	*file_out = file;
+	SC_FUNC_RETURN(ctx, SC_LOG_DEBUG_NORMAL, SC_SUCCESS);
+
+}
+
+static int cac_select_file(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
+{
+	return cac_select_file_by_type(card, in_path, file_out);
+}
+
+static int cac_finish(sc_card_t *card)
+{
+	cac_private_data_t * priv = CAC_DATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	if (priv) {
+		cac_free_private_data(priv);
+	}
+	return SC_SUCCESS;
+}
+
+
+/* select a CAC pki applet by index */
+static int cac_select_pki_applet(sc_card_t *card, int index)
+{
+	sc_path_t applet_path = cac_cac_pki_obj.path;
+	applet_path.aid.value[applet_path.aid.len-1] = index;
+	return cac_select_file_by_type(card, &applet_path, NULL);
+}
+
+/*
+ *  Find the first existing CAC applet. If none found, then this isn't a CAC
+ */
+static int cac_find_first_pki_applet(sc_card_t *card, int *index_out)
+{
+	int r, i;
+
+	for (i = 0; i < MAX_CAC_SLOTS; i++) {
+		r = cac_select_pki_applet(card, i);
+		if (r == SC_SUCCESS) {
+			u8 data[2];
+			sc_apdu_t apdu;
+
+			/* Try to read first two bytes of the buffer to
+			 * make sure it is not just malfunctioning card
+			 */
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_2,
+			    CAC_INS_GET_CERTIFICATE, 0x00, 0x00);
+			apdu.le = 0x02;
+			apdu.resplen = 2;
+			apdu.resp = data;
+			r = sc_transmit_apdu(card, &apdu);
+			/* SW1 = 0x63 means more data in CAC1 */
+			if (r == SC_SUCCESS && apdu.sw1 != 0x63)
+				continue;
+
+			*index_out = i;
+			return SC_SUCCESS;
+		}
+	}
+	return SC_ERROR_OBJECT_NOT_FOUND;
+}
+
+static int cac_populate_cac1(sc_card_t *card, int index, cac_private_data_t *priv)
+{
+	int r, i;
+	cac_object_t pki_obj = cac_cac_pki_obj;
+	u8 buf[100];
+	u8 *val;
+	size_t val_len;
+
+	/* populate PKI objects */
+	for (i = index; i < MAX_CAC_SLOTS; i++) {
+		r = cac_select_pki_applet(card, i);
+		if (r == SC_SUCCESS) {
+			pki_obj.name = get_cac_label(i);
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "CAC: pki_object found, cert_next=%d (%s),", i, pki_obj.name);
+			pki_obj.path.aid.value[pki_obj.path.aid.len-1] = i;
+			pki_obj.fd = i+1; /* don't use id of zero */
+			cac_add_object_to_list(&priv->pki_list, &pki_obj);
+		}
+	}
+
+	/*
+	 * create a cuid to simulate the cac 2 cuid.
+	 */
+	priv->cuid = cac_cac_cuid;
+	/* create a serial number by hashing the first 100 bytes of the
+	 * first certificate on the card */
+	r = cac_select_pki_applet(card, index);
+	if (r < 0) {
+		return r; /* shouldn't happen unless the card has been removed or is malfunctioning */
+	}
+	val = buf;
+	val_len = sizeof(buf);
+	r = cac_cac1_get_certificate(card, &val, &val_len);
+	if (r >= 0) {
+		priv->cac_id = malloc(20);
+		if (priv->cac_id == NULL) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+#ifdef ENABLE_OPENSSL
+		SHA1(val, val_len, priv->cac_id);
+		priv->cac_id_len = 20;
+		sc_debug_hex(card->ctx, SC_LOG_DEBUG_VERBOSE,
+		    "cuid", priv->cac_id, priv->cac_id_len);
```

Comment:
use tabs for indenting

Reply:
The indentation was changed in the last commit. I can squash them if the that is everything.